rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,610 @@
+revision of the resource in the update request matches the revision of the item in the backend. Conditional updates
+should also be preferred over traditional `CompareAndSwap` operations.
+
+```go
+func (s *FooService) UpdateFoo(ctx context.Context, foo *foov1.Foo) (*foov1.Foo, error) {
+    // The revision is cached prior to converting to the value because 
+    // conversion functions may set the revision to "" if MarshalConfig.PreserveResourceID
+    // is not set.
+    rev := foo.GetRevision()
+    value, err := convertFooToValue(foo)
+    if err != nil {
+        return nil, trace.Wrap(err)
+    }
+    item := backend.Item{
+        Key:     backend.Key("foo", foo.GetName()),
+        Value:   value,
+        Expires: foo.Expiry(),
+        Revision: rev,
+    }
+
+    lease, err := s.backend.ConditionalUpdate(ctx, item)
+    if err != nil {
+        return nil, trace.Wrap(err)
+    }
+
+    // Update the foo with the revision generated by the backend during the write operation.
+    foo.SetRevision(lease.Revision)
+    return foo, nil
+}
+```
+
+#### Upsert
+
+When upserting a resource, the `backend.Backend.Put` method should be used to persist the resource. It is also
+imperative that the revision generated by the backend is set on the returned resource.
+
+A resource may expose an upsert method from the backend layer even if the gRPC API does not expose an `Upsert` RPC. This
+may occur if a resource is cached, since the
+[cache collections](https://github.com/gravitational/teleport/blob/004d0db0c1f6e9b312d0b0e1330b6e5bf1ffef6e/lib/cache/collections.go#L60)
+require an upsert mechanism,
+see [`services.DynamicAccessExt`](https://github.com/gravitational/teleport/blob/004d0db0c1f6e9b312d0b0e1330b6e5bf1ffef6e/lib/services/access_request.go#L260-L278)
+for an example.
+
+```go
+func (s *FooService) UpsertFoo(ctx context.Context, foo *foov1.Foo) (*foov1.Foo, error) {
+    value, err := convertFooToValue(foo)
+    if err != nil {
+        return nil, trace.Wrap(err)
+    }
+    item := backend.Item{
+        Key:     backend.Key("foo", foo.GetName()),
+        Value:   value,
+        Expires: foo.Expiry(),
+    }
+
+    lease, err := s.backend.Put(ctx, item)
+    if err != nil {
+        return nil, trace.Wrap(err)
+    }
+
+    // Update the foo with the revision generated by the backend during the write operation.
+    foo.SetRevision(lease.Revision)
+    return foo, nil
+}
+```
+
+#### Get
+
+To retrieve a resource the `backend.Backend.Get` method should be provided a key built from the match parameters of the
+request.
+
+```go
+func (s *AccessService) GetFoo(ctx context.Context, name string) (*Foo, error) {
+    if name == "" {
+        return nil, trace.BadParameter("missing foo name")
+    }
+
+    item, err := s.backend.Get(ctx, backend.Key("foo", name))
+    if err != nil {
+        if trace.IsNotFound(err) {
+            return nil, trace.NotFound("foo %v is not found", name)
```

Comment:
I see. Could we push that into the backend, though? It seems laborious having this everywhere, plus I'm sure I'll forget it (and leak the key like you said).

Reply:
I was going to suggest that it could do an educated guess based on the last part of the key, but that isn't necessarily correct or friendly. Oh, well.